packages/runtime/src/decorators/middlewares.ts

```diff
@@ -0,0 +1,82 @@
+export type ExpressMiddleware = (req: any, res: any, next: any) => Promise<any>;
+export type KoaMiddleware = (ctx: any, next: any) => Promise<any>;
+export type HapiMiddlewareBase = (request: any, h: any) => Promise<any>;
+export type HapiMiddlewareSimple = HapiMiddlewareBase | { method: HapiMiddlewareBase; assign?: string; failAction?: HapiMiddlewareBase | string };
+export type HapiMiddleware = HapiMiddlewareSimple | HapiMiddlewareSimple[];
+
+export type Middlewares = {
+  express?: ExpressMiddleware[];
+  koa?: KoaMiddleware[];
+  hapi?: HapiMiddleware[];
+};
+
+/**
+ * Helper function to create a decorator
+ * that can act as a class and method decorator.
+ * @param fn a callback function that accepts
+ *           the subject of the decorator
+ *           either the constructor or the
+ *           method
+ * @returns
+ */
+function decorator(fn: (value: any) => void) {
+  return (...args: any[]) => {
+    // class decorator
+    if (args.length === 1) {
+      fn(args[0]);
+    } else if (args.length === 3 && args[2].value) {
+      // method decorator
+      const descriptor = args[2] as PropertyDescriptor;
+      if (descriptor.value) {
+        fn(descriptor.value);
+      }
+    }
+  };
+}
+
+/**
+ * Install `express`, `koa` and `hapi` middlewares
+ * to the Controller or a specific method.
+ * @param middlewares
+ * @returns
+ */
+export function Middlewares(middlewares: Middlewares) {
+  return decorator(target => {
+    target._expressMiddlewares = middlewares.express || [];
```

Comment:
I'd prefer using the reflect API here.
It seems like you're reimplementing some of the concepts that API offers manually in the template etc. Is there a reason why your approach is preferable that I'm missing?

Reply:
Hi, I'll try to take a look at the reflect API and let you know.
I'm not sure what I am reimplementing, if there's already a way to set custom middlewares for the various servers please let me know.
I should add that I'd really prefer to not use a custom routes template, and this feature seems to have been requested a number of times, so it seems to me that it's something the `tsoa` users would like.

Reply:
What this PR is doing is patching the target with some semi-transparent properties.
However, there's an API that works well and hides all of that patching: `Reflect.defineMetadata` and `Reflect.getMetadata`.

https://rbuckton.github.io/reflect-metadata/

Reply:
@WoH now I see what you mean, I didn't know about `reflect-metadata`, I updated the code to make use of it.